### PR TITLE
DM command

### PR DESCRIPTION
**Pull request by sjkd23**
/dm Command that allows higher ups to DM multiple users at once the same message, through the bot.

- Command works with and without command options.
        - Up to 15 users allowed with command options/shortcut
- Message limit of 1014 chars (embed field limit).
- If only message or only users were specified through command options, informs user that it does not work like that.
- If no message or users were specified in command options, sends embed with buttons, and will direct user to add a message and set all the members they'd like the bot to dm.

### CODE REVIEW

**Comment by sjkd23** [comment]
A couple of thoughts:

- Maybe a different command name? (like /dmall, or /sendmessage?)
- I'm not sure how DM rate limiting works for discord bots, so may need to set a message recipient cap.
- Has not been tested to DM more than one person (me).

File: src/commands/dm.ts [hunk 1]
```diff
@@ -0,0 +1,345 @@
+import { 
+ChatInputCommandInteraction,
```
Comment (ewang2002) [inline src/commands/dm.ts]: formatting

File: src/utils.ts [hunk 2]
```diff
@@ -40,4 +40,26 @@ export async function replyWithEmbed(embed: EmbedBuilder, interaction: CommandIn
         console.log(`ERROR: could not send embed in ${interaction.channelId}`);
         return;
     }
-}
\ No newline at end of file
+}
+
+export async function tryDelete(m: Message): Promise<boolean> {
```
Comment (ewang2002) [inline src/utils.ts:38]: docs

File: src/utils.ts [hunk 3]
```diff
@@ -40,4 +40,26 @@ export async function replyWithEmbed(embed: EmbedBuilder, interaction: CommandIn
...
+    try{
+        m.delete();
+        return true;
+    }catch(err){
+        console.error('could not delete message', err);
+        return false;
+    }
+}
+
+export async function tryToDMEmbed(embed: EmbedBuilder, member: User): Promise<boolean>{
```
Comment (ewang2002) [inline src/utils.ts:54]: docs

File: src/commands/dm.ts [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+            let DM = '';
+
+            buttonCollector.on('collect', async i => {
+
+                if (i.customId === 'add_message') {
+                    await i.update({ embeds: [addMsgEmbed], components: [addMsgRow] })
+
+                    msgCollector = interaction.channel!.createMessageCollector({
+                        filter: (m) => m.author.id === userID,
+                        time: 60_000,
```
Comment (ewang2002) [inline src/commands/dm.ts]: put into constant

File: src/commands/dm.ts [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+                    startEmbed.setTitle('DM User(s)')
+                        .setFields(
+                            { name: 'Please use the buttons below to: ', value: '- Create your desired message.' },
+                            { name: ' ', value: '- Choose which user(s) you would like the bot to send the message to.' },
+                            { name: ' ', value: '- Finally, click the "Send" button when you are finished.' }
+                        );
+                    if (memberList.length > 0) {
+                        startEmbed.addFields({ name: '__Message Recipients:__ ', value: memberList.join(', ') });
+                    }
+                    if (DM.length > 0 && DM !== '**__Message too long! (limit: 4096 characters)__**') {
```
Comment (ewang2002) [inline src/commands/dm.ts]: make this string a constant

File: src/commands/dm.ts [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+                    await sendDMs(DM, memberList);
+                }
+            });
+        }
+        /**
+         * function specifically for this command.
+         * Attempts to send specified message to all specified users.
+         * 
+         * @param {string} m - The message that is to be sent to all users 
+         * @param {User[]}users - Object array of all specified discord users.
```
Comment (ewang2002) [inline src/commands/dm.ts]:
```suggestion
         * @param {User[]} users - Object array of all specified discord users. 
```

File: src/commands/dm.ts [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+        /**
+         * function specifically for this command.
+         * Attempts to send specified message to all specified users.
+         * 
+         * @param {string} m - The message that is to be sent to all users 
+         * @param {User[]}users - Object array of all specified discord users. 
+         */
+        async function sendDMs(m: string, users: User[]): Promise<void> {
+            const DMEmbed = new EmbedBuilder()
+
```
Comment (ewang2002) [inline src/commands/dm.ts]: spacing

**Review by ewang2002 — CHANGES_REQUESTED** [review]

File: src/commands/dm.ts [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+                filter: i => i.user.id === interaction.user.id,
+                componentType: ComponentType.Button,
+                time: TIME_IN_MS.TEN_MINUTES
+            });
+
+            let msgCollector: MessageCollector;
+            let userCollector: MessageCollector;
+            let memberList: User[] = [];
+            let DM = '';
+            const msgTooLong = '**__Message too long! (limit: 1024 characters)__**';
```
Comment (ewang2002) [inline src/commands/dm.ts]: Put this outside of the function (like below the `MAX_USERS` const, with same naming convention, i.e., `MSG_TOO_LONG`)

File: src/commands/dm.ts [hunk 9]
```diff
@@ -0,0 +1,353 @@
...
+                .setComponents(BUTTONS.BACK_BUTTON, BUTTONS.DONE_BUTTON);
+
+            const addUsersRow = new ActionRowBuilder<ButtonBuilder>()
+                .setComponents(BUTTONS.BACK_BUTTON, BUTTONS.DONE_BUTTON, BUTTONS.RESET_BUTTON);
+
+            const confirmRow = new ActionRowBuilder<ButtonBuilder>()
+                .setComponents(BUTTONS.BACK_BUTTON, BUTTONS.YES_BUTTON);
+
+
+            const embedMessage = await interaction.reply({ content: 'Please follow the instructions below!', embeds: [startEmbed], components: [startRow] });
```
Comment (ewang2002) [inline src/commands/dm.ts]: This line is pretty long

File: src/commands/dm.ts [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+                    });
+
+                    msgCollector.on('collect', async m => {
+                        await tryDelete(m);
+                        if (m.content) {
+                            DM = m.content;
+                            if (DM.length > 1024) {
+                                DM = msgTooLong;
+                            }
+                            addMsgEmbed.setFields({ name: '__Current Message:__ ', value: DM });
```
Comment (ewang2002) [inline src/commands/dm.ts]: `'__Current Message:__ '` should be a constant :)

File: src/commands/dm.ts [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+                                }
+                            }
+                            if (member) {
+                                if (memberList.includes(member.user)) {
+                                    continue;
+                                } else {
+                                    memberList.push(member.user);
+                                }
+                            }
+                            addUsersEmbed.setFields({ name: '__Message Recipients:__ ', value: memberList.join(', ') });
```
Comment (ewang2002) [inline src/commands/dm.ts]: constant

File: src/commands/dm.ts [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+                        startEmbed.setFields(
+                            { name: '__Message Recipients:__ ', value: memberList.join(', ') },
+                            { name: '__Current Message:__ ', value: DM }
+                        );
+                        await interaction.editReply({ embeds: [startEmbed], components: [confirmRow] });
+                        await i.update({ components: [confirmRow] });
+                    }
+                } else if (i.customId === BUTTONS.YES_ID) {
+                    buttonCollector.stop();
+                    await i.update({})
```
Comment (ewang2002) [inline src/commands/dm.ts]: `deferUpdate` or `deferReply` (look up which one to use)

**Review by ewang2002 — APPROVED** [review]
approval given assuming you fix the things I pointed out

**Pull request merged**